### PR TITLE
Don't calculate incoming tx fee in byron

### PR DESCRIPTION
**Pull request by Unisay**
There is no point in calculating incoming Tx fees in Byron era as delta between total input values and total output values because not all inputs values are known.

Following https://github.com/input-output-hk/cardano-wallet/pull/3785

### Issue Number

ADP-2840

### CODE REVIEW

File: lib/wallet/src/Cardano/Wallet/Primitive/Model.hs [hunk 1]
```diff
@@ -791,13 +793,13 @@ applyOurTxToUTxO !slot !blockHeight !s !tx !u0 =
     -- NOTE 2: We do not have in practice the actual input amounts, yet we
     -- do make the assumption that if one input is ours, then all inputs are
     -- necessarily ours and therefore, known as part of our current UTxO.
-    actualFee direction = fromMaybe byronFee (tx ^. #fee)
+    actualFee direction = fromMaybe (Coin 0) (tx ^. #fee <|> byronFee)
```
Comment (piotr-iohk) [inline lib/wallet/src/Cardano/Wallet/Primitive/Model.hs:796]: I think it would be good to have a unit/property test for this function and/or capture this behavior in integration test. (I suppose we don't atm as previous change which also changed the behavior passed all tests)

**Review by HeinrichApfelmus — APPROVED** [review]
This code does what it is supposed to do. Looks good to me! 👍

File: lib/wallet/src/Cardano/Wallet/Primitive/Model.hs (hunk 1 above)
Comment (jonathanknowles) [reply lib/wallet/src/Cardano/Wallet/Primitive/Model.hs:796]: > I think it would be good to have a unit/property test for this function and/or capture this behavior in integration test. (I suppose we don't atm as previous change which also changed the behavior passed all tests)

@piotr-iohk One **_possible_** way would be to modify the various property tests for `applyBlocks` so that they verify that any fee transformation is consistent with our expectations.
 
 (Note that a **_previous_** attempt to do this was here: https://github.com/input-output-hk/cardano-wallet/pull/3801. This attempt simply asserted that fees were never transformed: of course, this expectation is now **_incorrect_**. Counterintuitively, the modified tests still pass even after merging this PR (#3806). This is because our transaction sequence generator never generates transactions with fees of `Nothing`, and therefore the fee modification code (which only kicks in **_if_** the fee is `Nothing`) is not actually covered by these (modified) properties. See this comment: https://github.com/input-output-hk/cardano-wallet/pull/3801#issuecomment-1480480749)

Comment (piotr-iohk) [reply lib/wallet/src/Cardano/Wallet/Primitive/Model.hs:796]: @jonathanknowles agreed.

@Unisay @HeinrichApfelmus what are the plans for testing this? The statement "code does what it is supposed to do" is quite weak as it is not backed up with any verification. I believe that if we make a change and we find out it is not covered with tests we should add sufficient testing.

Comment (HeinrichApfelmus) [reply lib/wallet/src/Cardano/Wallet/Primitive/Model.hs:796]: > "code does what it is supposed to do"

With this, I mean that I have inspected the code for `actualFee` and concluded that it does what I expect for different parameter cases, e.g. `direction = Outgoing` and `tx ^. #fee = Nothing`.

>  I believe that if we make a change and we find out it is not covered with tests we should add sufficient testing.

The issue is: Tests can only detect deviations of an implementation from a specification. What is the specification and what is the implementation?

In this instance, one can argue that the code for `actualFee` is simple and can serve as a specification, which propagates to the end user by composition.

* If we add a test case for `actualFee`, this would be the same as adding a second specification — which is not terribly useful for the given code, but could serve as a regression test, so that changing the specification requires changing both the code and the test.
* If we add a test case for `applyBlocks`, we would be attempting to test that both `actualFee` and its composition with the rest of the code is correct. However, for `actualFee`, the previous point applies, and for the composition, the structure of the code is such that I have little reason to doubt it beyond what is established by other tests, e.g. whether the definition of `totalIn` is correct.

I would be in favor of making `actualFee` a top-level function and adding some comments as to why the Byron case is treated differently, so that its status as _specification_ becomes more apparent. But apart from that, I see diminishing value in adding a second specification in the tests.

Comment (piotr-iohk) [reply lib/wallet/src/Cardano/Wallet/Primitive/Model.hs:796]: >The issue is: Tests can only detect deviations of an implementation from a specification. What is the specification and what is the implementation?

While it is true that tests can detect deviations of an implementation from a specification, it is not the only purpose they serve. Tests also help to ensure that the implementation is correct and that it works as intended.

Adding a test case for `actualFee` would not be the same as adding a second specification. It would be an additional layer of validation that helps to ensure that the implementation of the function is correct. This is particularly important in cases where the function is complex or has corner cases that may be difficult to catch by simply reading the code.

In the case of `applyBlocks`, testing the composition with the rest of the code is important because it helps to ensure that the function is working correctly in the context of the larger system. While the structure of the code may make it seem unnecessary to test the composition, it is still valuable to do so in order to catch any unexpected interactions between different parts of the code. The same point applies to integration testing btw.

Making `actualFee` a top-level function and adding comments may make its status as a specification more apparent, but it does not replace the need for testing. Comments can be helpful for understanding the code, but they do not provide the same level of validation as testing.

In summary, while tests cannot replace a specification, they are an important tool for ensuring that the implementation is correct and that it works as intended. Adding test cases for `actualFee` and `applyBlocks` would help to catch any issues that may be difficult to spot by simply reading the code or comments. 

Additionally, while the current code may not have any issues that are caught by the proposed tests, having them in place would be beneficial for the future in case of any changes to the code. Testing can help catch issues that may arise due to different or adjacent changes, making it a valuable tool for ensuring the ongoing reliability and correctness of the codebase.

Comment (HeinrichApfelmus) [reply lib/wallet/src/Cardano/Wallet/Primitive/Model.hs:796]: > While it is true that tests can detect deviations of an implementation from a specification, it is not the only purpose they serve. Tests also help to ensure that the implementation is correct and that it works as intended.

But "is correct" and "works as intended" are the same as "agrees with the specification". 😅 What does "correct" mean otherwise?

My argument is just that in this particular instance, the code for `actualFee` is of a form that it is its own specification (or at least can be brought into this shape — I would not object to `<|>` being replaced with a more basic pattern match if that matches the words of the specification better).

I mean, here the specification answers the question "What should the fee be?". And the answer in human words is "Well, the fee is part of the transaction data, unless the we are dealing with a Byron transaction, then in the case of `Incoming`, we set it to `0`, and in the case of `Outgoing`, we guess that it's the distance between incoming and outgoing funds for the wallet". The code for `actualFee` is a direct translation of these words — or rather, any translation of these words into a test will be of equal or worse directness.

> In the case of applyBlocks, testing the composition with the rest of the code is important because it helps to ensure that the function is working correctly in the context of the larger system.

As a pure function, `actualFee` is independent of the larger system. But I do agree that it could be supplied with wrong arguments ("garbage in, garbage out"), this is what I mean by "composition". Testing the composition would be about testing that the logic of the fee computation is exercised, and with the correct arguments (here: incoming and outgoing funds), but not about testing the results of this logic.

> Additionally, while the current code may not have any issues that are caught by the proposed tests, having them in place would be beneficial for the future in case of any changes to the code. 

This would be a regression test, whose purpose is to protect against "accidental" changes to the code. This protection works like this: a change of the code leads to a test failure — indicated that code and its specification no longer match, where the test file serves as a specification. Then, the programmer has a choice: Change the code to match the specification, or change the specification to match the code, depending on which is perceived as "higher correctness status". If `actualFee` is a local helper function, I can see that it doesn't score high on the "correctness status" and the programmer may be tempted to change it without thinking. However, if `actualFee` is a top-level function, then it's perceived "correctness status" is much higher. Being in a separate file, the test also has a high "correctness status", but duplicating `actualFee` into such a test file in order to elevate its perceived status seems dubious to me.

I suppose my argument boils down to this:

* Tests are of high value if they approach the code from a different angle and do not duplicate it. Unfortunately, that is not the case for `actualFee`.

Comment (piotr-iohk) [reply lib/wallet/src/Cardano/Wallet/Primitive/Model.hs:796]: >Tests are of high value if they approach the code from a different angle and do not duplicate it. Unfortunately, that is not the case for actualFee.

Thank you for sharing your perspective on this. I understand that you feel that the implementation of the `actualFee` function is so simple and self-contained that it serves as its own specification. However, I still believe that having tests in place to ensure that the function is working correctly in all scenarios is valuable.

While it is true that the specification is the ultimate measure of correctness, having tests that exercise the function and catch any unexpected issues can help ensure that the implementation is actually meeting the specification (or at least validating our own expectation to what the given function/piece od the code actually does). In addition, as you mentioned, having tests in place can serve as regression tests to catch any accidental changes to the code in the future. 

**Case in point**... as far as I see on current master (`v2022-12-14 (git revision: 6f49a1423d4ccb4152e5d6d3e61bde21175c46d7)`) the behavior is not as intended. The fee is still displayed (>0) on incoming transactions for Byron wallets. Needless to say that this would have been immediatelly caught if the change here was supplied with sufficient testing. :(

Comment (HeinrichApfelmus) [reply lib/wallet/src/Cardano/Wallet/Primitive/Model.hs:796]: > The fee is still displayed (>0) on incoming transactions for Byron wallets. Needless to say that this would have been immediately caught if the change here was supplied with sufficient testing. :(

Is that on a fresh wallet, or a wallet that has synchronized with a previous version of `master` where the issue ("fee `>0` for incoming Byron transaction") was present? 🤔

For the latter case: The `fee` is written to the database, and the state produced by a buggy version of `master` will persist to a new version of `master`. 😅 … and this won't be caught by an integration test running on a fresh cluster. 😅😅

Comment (piotr-iohk) [reply lib/wallet/src/Cardano/Wallet/Primitive/Model.hs:796]: It is on wallet that is restored from scratch on `master`. :man_shrugging:

Comment (HeinrichApfelmus) [reply lib/wallet/src/Cardano/Wallet/Primitive/Model.hs:796]: > It is on wallet that is restored from scratch on master. 🤷‍♂️

Understood. 🤔 I'd like to clarify the intended behavior — do you mean "Byron wallet" or "Byron transaction"? It would consider it ok for a Byron wallet to show `> 0` fee on "Shelley transactions", i.e. transactions that occur in the Shelley era or later. (The technical reason is that as of that era, transactions fees are officially part of the transaction data and are easy to display.)

Comment (piotr-iohk) [reply lib/wallet/src/Cardano/Wallet/Primitive/Model.hs:796]: I mean Byron wallet. The behavior is:
 - if you send a transaction from either Byron or Shelley wallet to a Byron wallet
 - fee on `outgoing` tx from source wallet > 0 (which is expected)
 - fee on `incoming` tx on target Byron wallet > 0 (which I thought is not expected now and all fees on incoming txs on Byron wallets should be displayed as `0`)

TBH, the intended behavior is now very confusing to me. If the change applies only to transactions that could be done in the Byron era, then I don't see the point of it. We are long past Byron era. Unless there's something else I'm missing?

Comment (HeinrichApfelmus) [reply lib/wallet/src/Cardano/Wallet/Primitive/Model.hs:796]: > If the change applies only to transactions that could be done in the Byron era

Yes.

> then I don't see the point of it. We are long past Byron era. Unless there's something else I'm missing?

Well, it does affect how fees are shown for historical transactions from the Byron era — a few select users may have that in their wallet… 🤷

Presenting the `distance ingoing outgoing` as fee is wrong in this special case if the historical transactions was incoming — displaying `0` makes more sense.
